### PR TITLE
[StructuredAuthorizationConfiguration] Add --authorization-config flag and guard it using a Feature Gate

### LINKED ISSUE #118872: [StructuredAuthorizationConfig] - Implement AuthorizationConfiguration
Add a configuration format with specific precedence order and defined failure modes for configuring authorizer chain.

Implements https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration

- [x] create API types for config file, loading helper https://github.com/kubernetes/kubernetes/pull/119099
- [x] refactor existing authz structs to transform to authz config API type https://github.com/kubernetes/kubernetes/pull/119099
- [ ] add kube-apiserver flag to accept the configuration file https://github.com/kubernetes/kubernetes/pull/120154

xref: https://github.com/liggitt/kubernetes/commits/authz-config

/triage accepted
/sig auth

### PR DESCRIPTION
**Pull request by palnabarun**
#### What type of PR is this?

/kind feature
/area apiserver

#### What this PR does / why we need it:

This is follow-up on https://github.com/kubernetes/kubernetes/pull/119099 and implements the following:

- kube-apiserver flag (`--authorization-config`) for authorization configuration file
- feature gate (`StructuredAuthorizationConfiguration`) for gating changes
- adds an integration test for reading authorizers from the configuration file

#### Which issue(s) this PR fixes:

Fixes #118872

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
kube-apiserver: adds --authorization-config flag for reading a configuration file containing an apiserver.config.k8s.io/v1alpha1 AuthorizationConfiguration object. --authorization-config flag is mutually exclusive with --authorization-modes and --authorization-webhook-* flags. The alpha StructuredAuthorizationConfiguration feature flag must be enabled for --authorization-config to be specified.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://kep.k8s.io/3221
- [Usage]: TBD
- [Other doc]: TBD
```

### CODE REVIEW

**Comment by jiahuif** [comment]
/retitle [WIP][StructuredAuthorizationConfiguration] Add --authorization-config flag and guard it using a Feature Gate 
Please unmark WIP when it is ready to review. Thank you

**Comment by hailkomputer** [comment]
Hello! Bug triage team lead here. I want to check the status. 
The code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03) (about two weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged and issues are addressed on time.
As the PR is targeting 1.29, is it still planned for this release?

**Comment by palnabarun** [comment]
> As the PR is targeting 1.29, is it still planned for this release?
@hailkomputer yup, it is.

File: pkg/controlplane/apiserver/options/options_test.go [hunk 1]
```diff
@@ -283,6 +283,11 @@ func TestAddFlags(t *testing.T) {
 	expected.Authentication.OIDC.UsernameClaim = "sub"
 	expected.Authentication.OIDC.SigningAlgs = []string{"RS256"}
 
+	if !s.Authorization.AreLegacyFlagsSet() {
+		t.Errorf("expected legacy authorization flags to be set")
+	}
+	s.Authorization.AreLegacyFlagsSet = nil
```
Comment (liggitt) [inline pkg/controlplane/apiserver/options/options_test.go:289]: nit: add a comment to explain we nil this so the compare below works

File: pkg/kubeapiserver/options/authorization.go [hunk 2]
```diff
@@ -50,6 +65,10 @@ type BuiltInAuthorizationOptions struct {
 	// This allows us to configure the sleep time at each iteration and the maximum number of retries allowed
 	// before we fail the webhook call in order to limit the fan out that ensues when the system is degraded.
 	WebhookRetryBackoff *wait.Backoff
+
+	AuthorizationConfigurationFile string
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authorization.go:69]: add a comment about which fields this is mutually exclusive with

File: pkg/kubeapiserver/options/authorization.go [hunk 3]
```diff
@@ -111,27 +164,43 @@ func (o *BuiltInAuthorizationOptions) AddFlags(fs *pflag.FlagSet) {
...
+		authorizationWebhookUnauthorizedTTLFlag, o.WebhookCacheUnauthorizedTTL,
 		"The duration to cache 'unauthorized' responses from the webhook authorizer.")
+
+	fs.StringVar(&o.AuthorizationConfigurationFile, authorizationConfigFlag, o.AuthorizationConfigurationFile, ""+
+		"File with Authorization Configuration to configure the authorizer chain."+
+		"Note: This feature is in Alpha since v1.29."+
+		"--feature-gate=StructuredAuthorizationConfiguration=true feature flag needs to be set to true for enabling the functionality."+
+		"This feature is mutually exclusive with the other --authorization-mode and --authorization-webhook-* flags.")
+
+	if o.AreLegacyFlagsSet == nil {
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authorization.go:195]: if we already have a non-nil function, OR it

```go
oldLegacyFlagsSet := o.AreLegacyFlagsSet
o.AreLegacyFlagsSet = func() bool {
  if oldLegacyFlagsSet != nil && oldLegacyFlagsSet() {
    return true
  }
  return fs.Changed(authorizationModeFlag) ||
				fs.Changed(authorizationWebhookConfigFileFlag) ||
				fs.Changed(authorizationWebhookVersionFlag) ||
				fs.Changed(authorizationWebhookAuthorizedTTLFlag) ||
				fs.Changed(authorizationWebhookUnauthorizedTTLFlag)
}
```

File: pkg/kubeapiserver/options/authorization_test.go [hunk 4]
```diff
@@ -173,6 +173,12 @@ func TestBuiltInAuthorizationOptionsAddFlags(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	if !opts.AreLegacyFlagsSet() {
+		t.Fatal("legacy flags should have been configured")
+	}
+
+	opts.AreLegacyFlagsSet = nil
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authorization_test.go:180]: comment why we're clearing this

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go [hunk 5]
```diff
@@ -215,7 +215,7 @@ const (
 	TypeWebhook                                      AuthorizerType = "Webhook"
 	FailurePolicyNoOpinion                           string         = "NoOpinion"
 	FailurePolicyDeny                                string         = "Deny"
-	AuthorizationWebhookConnectionInfoTypeKubeConfig string         = "KubeConfigFile"
+	AuthorizationWebhookConnectionInfoTypeKubeConfig string         = "KubeConfig"
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:218]: drop out when rebased on the name fixup PR

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+func init() {
+	install.Install(scheme)
+}
+
+func LoadFromFile(file string) (*api.AuthorizationConfiguration, error) {
+	if len(file) == 0 {
+		// no file specified, use default config
+		return LoadFromData(nil)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go:43]: return an error here (or just let ReadFile return an error trying to read ""), we don't need to copy this pattern of using defaults from PSA admission

File: pkg/kubeapiserver/options/authorization.go [hunk 7]
```diff
@@ -69,6 +88,40 @@ func (o *BuiltInAuthorizationOptions) Validate() []error {
...
+		}
+
+		// validate the file and return any error
+		if errors := validation.ValidateAuthorizationConfiguration(nil, config,
+			sets.NewString(authzmodes.AuthorizationModeChoices...),
+			sets.NewString(repeatableAuthorizerTypes...),
+		); len(errors) != 0 {
+			allErrors = append(allErrors, errors.ToAggregate().Errors()...)
+		}
+
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authorization.go:120]: after calling generic validation, also ensure names of authorizers other than type=Webhook are equal to `lower(type)`

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 8]
```diff
@@ -305,6 +312,8 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	StructuredAuthenticationConfiguration: {Default: false, PreRelease: featuregate.Alpha},
 
+	StructuredAuthorizationConfiguration: {Default: false, PreRelease: featuregate.Alpha},
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:315]: also register this into pkg/features/kube_features.go with the other feature gates we inherit from genericfeatures

**Comment by liggitt** [comment]
there's a basic integration test at https://github.com/liggitt/kubernetes/commits/authz-config-external-changes you can pull in

**Comment by palnabarun** [comment]
@liggitt this should be good for a final review. I have pulled the integration test commit from your fork.

**Comment by liggitt** [comment]
/retest
(infra pod scheduling errors)

**Comment by liggitt** [comment]
/retest
unrelated etcd error in integration test

File: pkg/kubeapiserver/options/authorization.go [hunk 9]
```diff
@@ -69,6 +94,54 @@ func (o *BuiltInAuthorizationOptions) Validate() []error {
...
+		// error out if legacy flags are defined
+		if o.AreLegacyFlagsSet != nil && o.AreLegacyFlagsSet() {
+			return append(allErrors, fmt.Errorf("--%s can not be specified when --%s or --authorization-webhook-* flags are defined", authorizationConfigFlag, authorizationModeFlag))
+		}
+
+		// load the file and check for errors
+		config, err := load.LoadFromFile(o.AuthorizationConfigurationFile)
+		if err != nil {
+			return append(allErrors, fmt.Errorf("failed to load AuthorizationConfiguration from file: %s", err))
```
Comment (ritazh) [inline pkg/kubeapiserver/options/authorization.go:116]:
```suggestion
			return append(allErrors, fmt.Errorf("failed to load AuthorizationConfiguration from file: %v", err))
```

File: pkg/kubeapiserver/options/authorization.go [hunk 10]
```diff
@@ -140,17 +233,52 @@ func (o *BuiltInAuthorizationOptions) ToAuthorizationConfig(versionedInformerFac
...
+
+		// error out if legacy flags are defined
+		if o.AreLegacyFlagsSet != nil && o.AreLegacyFlagsSet() {
+			return nil, fmt.Errorf("--%s can not be specified when --%s or --authorization-webhook-* flags are defined", authorizationConfigFlag, authorizationModeFlag)
+		}
+
+		// load the file and check for errors
+		authorizationConfiguration, err = load.LoadFromFile(o.AuthorizationConfigurationFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load AuthorizationConfiguration from file: %s", err)
```
Comment (ritazh) [inline pkg/kubeapiserver/options/authorization.go:259]:
```suggestion
			return nil, fmt.Errorf("failed to load AuthorizationConfiguration from file: %v", err)
```

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package load
+
+import (
+	"fmt"
+	"io"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
```
Comment (ritazh) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go:22]: should this be moved down next to all other k8s import?

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+func LoadFromFile(file string) (*api.AuthorizationConfiguration, error) {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	return LoadFromData(data)
+}
+
+func LoadFromReader(reader io.Reader) (*api.AuthorizationConfiguration, error) {
```
Comment (ritazh) [inline staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go:48]: where do we use this other than the test?

**Comment by liggitt** [comment]
ah, needs rebase

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go (hunk 12 above)
Comment (palnabarun) [reply staging/src/k8s.io/apiserver/pkg/apis/apiserver/load/load.go:48]: Only in test for now.

**Comment by palnabarun** [comment]
This PR shouldn't have the `needs-rebase` label anymore after changes were made hours ago.

**Comment by palnabarun** [comment]
What are the guidelines for manually removing the `needs-rebase` label?

**Comment by sftim** [comment]
Changelog suggestion
```markdown
kube-apiserver: added `--authorization-config` command line argument for reading a configuration file containing a serialized apiserver.config.k8s.io/v1alpha1 AuthorizationConfiguration. The `--authorization-config` argument is mutually exclusive with `--authorization-modes` and the `--authorization-webhook-*` arguments. The (alpha) `StructuredAuthorizationConfiguration` feature gate must be enabled for `--authorization-config` to be specified.
```

- most of the IT world uses “flag” to refer to a boolean. In Golang a command line flag can be anything, but this confuses people used to the more widely used convention.
- we call feature gates [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/); these are actually booleans but we still should be using the term that end users see

**Pull request merged**